### ISSUE TITLE
AppShells and Layouts

### ISSUE THREAD

**Issue by niktek** [Brain-Bones/skeleton#158]
Will be fleshing this out more - but with AppShells on the roadmap and also needing to revise my own site, I ran into the issue of svelte layouts not being able to have named slots as per https://github.com/sveltejs/kit/issues/627 which also has implications for Skeleton offering a nice OOB DX.

**Comment by endigo9740** [comment]
Hmm that's a bummer. Hopefully that's something they address. It would be hugely beneficial for this use case.

FYI, see the Roadmap > Docs > Doc Refresh section here:
https://github.com/Brain-Bones/skeleton/wiki/%F0%9F%9B%A3%EF%B8%8F-The-Skeleton-Roadmap-(proposal)

Part of that first bullet is "improved layout". I was kind of expecting to handle setup for the Nav Bar and App Shell during that process. I essentially want to try and "dogfood" ever component we can within the Docs site before we release it. So we know it works well in a production environment.

Given the issue you linked and the placement on the roadmap proposal, I'd say this one may be a ways out. But hopefully the named slot stuff is sorted by then! It may all work out.

**Comment by endigo9740** [comment]
The alternative, however, would be to treat the App Shell like a component, rather than a layout for now. You have to import and set it up on each page, but that may still be better than just rebuilding page layouts for scratch over and over. I'll think this about this!

**Comment by endigo9740** [comment]
Linking this for reference later. I think something like this might be helpful for generating the layout structure:
https://css-tricks.com/using-grid-named-areas-to-visualize-and-reference-your-layout/

**Comment by endigo9740** [comment]
FYI to use SvelteKit's routing group layouts I needed to update dependencies:
https://kit.svelte.dev/docs/advanced-routing#advanced-layouts-group

![Screen Shot 2022-08-29 at 1 48 06 PM](https://user-images.githubusercontent.com/1509726/187275538-1ad6de1e-4914-4406-902b-f21ff8b05b1b.png)

**Comment by endigo9740** [comment]
Great progress made on this today!

Notable improvements:

- We now have a dedicated AppBar (read: top nav bar) component
- We now have an AppShell component that houses left/right panels and the middle content
- The homepage has a unique layout and design, similar to Vite, Vitest, Tailwind, etc.
- The root layout now only sets the header/content area - it's MUCH cleaner than before
- I've setup a (inner) Route Group with it's own layout, this implements the AppShell component
- We now have a `/libs/_documentation` section for doc-specific components.
- These doc-specific components include an AppBar, Drawer, Footer, and ThemeGeneration components

When I return to this I'll look to flush out the homepage a bit more. There's a couple links that need to be fixed, and I need to set it up to fetch the list of contributors and their avatars. GitHub provides an API for this, but I'll need to explore it a bit deeper - needs an auth token, etc.

![homepage-1](https://user-images.githubusercontent.com/1509726/187312510-9407cf0c-a468-4351-a8f4-49432e74c873.png)

![homepage-2](https://user-images.githubusercontent.com/1509726/187312516-36729e90-7726-4dc9-bacc-9724a6108506.png)

![inner-page](https://user-images.githubusercontent.com/1509726/187312518-0478724c-87e6-4c1d-9c14-34aa98b43bca.png)

**Comment by endigo9740** [comment]
Here's a video showing the current layout for the homepage versus inner pages:
https://i.imgur.com/dReerEo.mp4

General idea is the AppShell provides a ton of potential slots:

> NOTE: the sidebars (in red) will most likely have a default width of 0px. This allows them to easily be toggled on/off based on conditions. Which is great for our homepage, where we don't want them visible, but still want the Drawer menu to be in scope to operate when on mobile.

![Screen Shot 2022-08-30 at 12 42 19 PM](https://user-images.githubusercontent.com/1509726/187507293-51fb90d1-b895-411f-bb19-7aab368da17d.png)

Here's the root layout markup:

![Screen Shot 2022-08-30 at 12 41 48 PM](https://user-images.githubusercontent.com/1509726/187507714-53e120fb-a9d7-4a79-9631-aa05dd25faf6.png)

Versus the (inner) route group layout:

![Screen Shot 2022-08-30 at 12 49 14 PM](https://user-images.githubusercontent.com/1509726/187507864-154eeb68-2571-4c48-9b68-6d962ee598d3.png)

**Comment by endigo9740** [comment]
Documentation added for the App Shell component. The props are pretty minimal on this one, but we may find cases where it'll be beneficial to add more.

I'd welcome suggestions if you have any @niktek .

![image](https://user-images.githubusercontent.com/1509726/187752754-a79010aa-9a9a-45ce-ad01-6f2efe53ece2.png)

**Comment by endigo9740** [comment]
Vitest integration cases have been added. I think with this the component is ready for QA!

Please pull branch `feat/app-shell` to begin testing asap @thomasbjespersen @niktek

**Comment by endigo9740** [comment]
Re-opening this so QA can continue.

**Comment by endigo9740** [comment]
FYI I constructed a simple Sveltekit app and implemented the AppShell, AppBar, and Drawer components. Everything worked as expected following the full documentation install instructions plus a local .tgz tarball version of the Skeleton package.

![Screen Shot 2022-09-05 at 3 25 18 PM](https://user-images.githubusercontent.com/1509726/188509640-30723d67-c3ce-4185-b7ae-c4dc64778664.png)


However I did run into two small issues...

First, an ARIA Heading error log in the terminal when running `npm run test`:

```
/.../node_modules/@brainandbones/skeleton/GradientHeading/GradientHeading.svelte:11:99 A11y: Elements with the ARIA role "heading" must have the following attributes defined: "aria-level"
```

Second, I noted the Drawer "fly" animation was missing to make it appear to slide out from the edge. Compare the gifs of the Docs site animation version the test app using the package:
https://imgur.com/a/w7NBnp9
